File: comms/native/services/validation/DeltaManager.js
// Copyright: IQGeo Limited 2010-2023

import myw, { MywClass, MywInternalError, FilterParser } from 'myWorld-base';
import { Conflict, FeatureChange } from 'myWorld-native-services';
import NetworkView from '../api/NetworkView';
import DataValidator from './DataValidator';

/**
 * Engine for managing design data
 */
/*eslint-disable no-await-in-loop*/
class DeltaManager extends MywClass {
    /**
     * Init slots of self
     *
     * DB_VIEW is a FeatureView
     */
    constructor(db_view, progress) {
        super();
        if (!db_view.delta) {
            throw new MywInternalError('View not delta', db_view); // TODO: Class not defined
        }

        this.db_view = db_view;
        this.progress = progress;

        this.nw_view = new NetworkView(this.db_view, this.progress);
        // this.conflict_fixer = new ConflictResolver(this.nw_view,this.progress)  // Not supported in native
        // this.data_fixer     = new DataFixer(this.nw_view,this.progress)        // Not supported in native
    }

    // ------------------------------------------------------------------------------
    //                                   CHECKING
    // ------------------------------------------------------------------------------

    /**
     * The feature changes made in self's delta
     *
     * Returns an ordered list of FeatureChange objects
     */
    async changes(
        feature_types = undefined,
        change_types = undefined,
        bounds = undefined,
        limit = undefined
    ) {
        if (change_types === undefined) {
            change_types = ['insert', 'update', 'delete'];
        }

        const changes = [];
        const delta_recs = await this.deltaRecs(feature_types, change_types, bounds);
        let cnt = 0;
        for (const delta_rec of delta_recs) {
            if (limit && cnt >= limit) break;
            cnt += 1;

            let base_rec = undefined;
            if (delta_rec.myw.change_type == 'update') {
                const table = await this.db_view.table(delta_rec.getType());
                base_rec = await table._baseRec(delta_rec.id);
                base_rec.view = this.db_view;
            }

            const change = new FeatureChange(delta_rec.myw.change_type, delta_rec, base_rec);
            changes.push(change);
        }

        return changes;
    }

    /**
     * Find conflicts in current delta
     *
     * Optional BOUNDS is a polygon. Optional CATEGORIES is a list of names
     *
     * Returns a list of lists of MywConflict objects, keyed by feature type
     */
    async conflicts(bounds = undefined, categories = undefined) {
        const feature_types = await this.orderedFeatureTypes(categories);
        const change_types = ['insert', 'update', 'delete'];

        // Find conflicts
        const conflicts = {};

        for (const feature_type of feature_types) {
            const table = await this.db_view.table(feature_type);

            // Find changed records
            const delta_recs = await this.deltaRecs([feature_type], change_types, bounds);

            // Find conflicts
            const ft_conflicts = {};
            for (const delta_rec of delta_recs) {
                const conflict = await this.conflictFor(table, delta_rec);

                if (conflict) {
                    delta_rec.view = this.db_view;
                    ft_conflicts[delta_rec.id] = conflict;
                }
            }

            // Add to result
            if (Object.keys(ft_conflicts).length) {
                conflicts[feature_type] = ft_conflicts;
            }
        }

        return conflicts;
    }

    /**
     * Check integrity of objects in self's delta
     *
     * Returns a list of IntegrityError objects
     *
     * MAX_ERRORS - if specified will stop after that many errors found
     */
    async validate(bounds = undefined, categories = undefined, max_errors = undefined) {
        this.progress(6, 'Checking', this.db_view);
        const validator = new DataValidator(this.db_view, undefined, this.progress);

        const feature_types = await this.orderedFeatureTypes(categories);

        const delta_recs = await this.deltaRecs(feature_types, undefined, bounds);
        for (const rec of delta_recs) {
            await validator.check(rec);

            if (max_errors && Object.keys(validator.errors).length >= max_errors) {
                break;
            }
        }

        return validator.errors;
    }

    // ------------------------------------------------------------------------------
    //                                  MERGING
    // ------------------------------------------------------------------------------
    // Not supported on native
    // /**
    //  * Auto-resolve conflicts and integrity errors in self's delta
    //  *
    //  * Returns a list of FeatureChange objects
    //  */
    // merge() {
    //     this.progress(2, 'Merging', this.db_view);

    //     changes = {};
    //     this.fixConflicts(changes);
    //     this.fixGeoms(changes);

    //     return changes.values();
    // },

    // /**
    //  * Auto-resolve conflicts in self's delta
    //  *
    //  * Updates CHANGES (a set of FeatureChange objects)
    //  */
    // fixConflicts(changes) {
    //     this.progress('Fixing conflicts in', this.db_view);

    //     for (const [feature_type, conflicts] of this.conflicts().items()) {
    //         for (const [id, conflict] of conflicts.items()) {
    //             this.fixConflict(conflict, changes);
    //         }
    //     }
    // },

    // /**
    //  * Auto-resolve CONFLICT (if possible)
    //  *
    //  * Returns a FeatureChange (None if no change)
    //  */
    // fixConflict(conflict, changes) {
    //     conflict_type = conflict.master_change + '/' + conflict.delta_rec.myw_change_type;

    //     // Say what we are doing
    //     this.progress(6, conflict.delta_rec, 'Attempting to auto-resolve', conflict_type);

    //     // Resolve
    //     if (conflict_type == 'update/update')
    //         return this.fixUpdateUpdateConflict(conflict, changes);
    //     if (conflict_type == 'update/delete')
    //         return this.fixUpdateDeleteConflict(conflict, changes);
    //     return undefined;
    // },

    /**
     * Return delta_rec to master state
     */
    // ENH: Don't support this in Native
    async revert(delta_rec) {
        // return this.conflict_fixer.revert(delta_rec);

        const table = await this.db_view.table(delta_rec.getType());

        if (!table.deltaTable) return;
        await table.deltaTable.delete(delta_rec.id);

        const base_rec = await table._baseRec(delta_rec.id);
        if (base_rec) await table.baseTable.delete(base_rec.id);
    }

    /**
     * No-op on native
     * @param {MywFeature} delta_rec
     * @returns
     */
    async rebase(delta_rec) {
        return delta_rec;
    }

    // ------------------------------------------------------------------------------
    //                                  HELPERS
    // ------------------------------------------------------------------------------

    /**
     * The delta records of current view
     */
    // ENH: Replace by protocol on MywVersionedFeatureView
    // ENH@ Return detached records?
    async deltaRecs(feature_types = undefined, change_types = undefined, bounds = undefined) {
        if (change_types === undefined) {
            change_types = ['insert', 'update'];
        }

        // Deal with defaults
        if (feature_types === undefined) {
            feature_types = await this.orderedFeatureTypes();
        }

        // For each feature type ..
        const recs = [];
        for (const feature_type of feature_types) {
            const tab = await this.db_view.table(feature_type);
            await tab.initialized;
            // Build query (ordering to get inserts first)
            let query = tab._delta_recs.orderBy('id', false);
            if (change_types) {
                const change_type_strs = change_types.map(ct => `'${ct}'`);
                const filter = `[myw_change_type] in (${change_type_strs.join(',')})`;
                const pred = new FilterParser(filter).parse();
                query = query.filter([pred]);
            }

            if (bounds) {
                const primary_geom_name = tab.descriptor?.primary_geom_name;

                if (primary_geom_name) {
                    query = query.whereIntersects(primary_geom_name, bounds);
                }
            }

            // Yield records
            const ft_recs = await query.all();
            for (const rec of ft_recs) {
                recs.push(rec);
            }
        }

        return recs;
    }

    /**
     * Versioned feature types, in 'top down' order
     *
     * If optional CATEGORIES is provided, return only types from those categories
     */
    async orderedFeatureTypes(categories = undefined) {
        // ENH: Move to network view

        // Mapping from category name -> nw_view property
        const name_mappings = { structures: 'structs', equipment: 'equips' };

        // Case: Categories
        // ENH: Filter result instead
        if (categories) {
            let feature_types = [];
            for (const category of categories) {
                const prop_name = name_mappings[category] || category;
                let cat_fts = await this.nw_view[prop_name];
                if (!cat_fts.length) cat_fts = Object.keys(cat_fts); // ENH: Fix network_view categories to all be arrays
                feature_types = [...feature_types, ...cat_fts];
            }

            return feature_types;
        }

        // Case: All
        const all_feature_types = await this.getVersionedFeatureTypes();

        const user_feature_types = [
            ...Object.keys(this.nw_view.structs),
            ...Object.keys(this.nw_view.routes),
            ...Object.keys(this.nw_view.equips),
            ...Object.keys(this.nw_view.conduits),
            ...Object.keys(this.nw_view.cables),
            ...Object.keys(this.nw_view.connections),
            ...Object.keys(this.nw_view.circuits),
            ...Object.values(this.nw_view.line_of_counts)
        ];

        const int_feature_types = [
            ...Object.values(this.nw_view.conduit_runs),
            ...Object.keys(this.nw_view.segments)
        ];

        const custom_feature_types = [
            ...new Set(
                all_feature_types.filter(
                    feature =>
                        !user_feature_types.includes(feature) &&
                        !int_feature_types.includes(feature)
                )
            )
        ].sort();

        return [...custom_feature_types, ...user_feature_types, ...int_feature_types];
    }

    /**
     * Conflict info for delta_rec (if any)
     *
     * Returns a Conflict (or None)
     */
    async conflictFor(table, delta_rec) {
        // ENH: Remove need for custom conflict class and get rid of this

        const conflict = await table.conflictFor(delta_rec);

        if (!conflict) {
            return undefined;
        }

        return new Conflict(
            conflict.master_change,
            conflict.delta_rec,
            conflict.master_rec,
            conflict.base_rec
        );
    }

    /**
     * Returns geometry bounds for changes in delta
     * @returns Object with geometry property is bounds geometry or None
     */
    async bounds() {
        const delta = this.db_view.delta;
        const buffer_dist = 10;

        // Build query SQL
        const indexTables = [
            'myw$delta_geo_world_point',
            'myw$delta_geo_world_linestring',
            'myw$delta_geo_world_polygon'
        ];
        const indexSql = indexTables.map(
            table => ` select the_geom from ${table} where delta = '${delta}' `
        );
        const unionSql = indexSql.join(' UNION ');
        const aggSql = `AsGeoJSON(ST_Collect(the_geom))`;
        const sql = `select ${aggSql} as the_geom from ( ${unionSql} )`;

        const result = await this.db_view.db.runSql(sql);
        const geom = JSON.parse(result[0]['the_geom']);

        if (!geom) return { geometry: null };

        // Use turf to create the correct buffer. This can probably be done in SQL but more complex.
        /*eslint-disable no-undef*/
        await myw.geometry.init();
        const feature = { type: 'Feature', properties: {}, geometry: geom };
        const convex = turf.convex(feature);
        const buffer = turf.buffer(convex, buffer_dist, { units: 'meters' });
        /*eslint-enable no-undef*/

        return { geometry: buffer.geometry };
    }

    /**
     * TBR: PLAT-8203 When platform provides this method on the native server api
     * This is a copy of a method on NetworkView which is a copy of a method on WorkflowDeltaController
     * Gets all feature_types configured as versioned from the myWorld datasource=
     * @returns {Promise<Array>} an array of feature types
     */
    async getVersionedFeatureTypes() {
        const records = await this.db_view.db
            .cachedTable('dd_feature')
            .where({ datasource_name: 'myworld', versioned: true })
            .all();

        return records.map(rec => rec.feature_name);
    }
}

export default DeltaManager;

File: comms/server/validation/delta_manager.py
# Copyright: IQGeo Limited 2010-2023
from sqlalchemy.orm import Query
from sqlalchemy import func, literal

from myworldapp.core.server.base.core.myw_error import MywError, MywInternalError
from myworldapp.core.server.base.core.myw_progress import MywProgressHandler
from myworldapp.modules.comms.server.api.network_view import NetworkView
from myworldapp.core.server.base.geom.myw_geo_utils import degrees_to_metres
from myworldapp.core.server.base.geom.myw_geometry import MywGeometry
from myworldapp.core.server.base.db.globals import Session

# from myworldapp.core.server.controllers.myw_select_controller import getModels

from .data_validator import DataValidator
from .data_fixer import DataFixer
from .conflict_resolver import ConflictResolver
from .conflict import Conflict
from .feature_change import FeatureChange


class DeltaManager:
    """
    Engine for managing design data

    Provides API for getting changes, validating and merging a delta.
    Returns changes as FeatureChange objects"""

    def __init__(self, db_view, progress=MywProgressHandler()):
        """
        Init slots of self

        DB_VIEW is a MywFeatureView"""

        if not db_view.delta:
            raise MywInternalError("View not delta", db_view)

        self.db_view = db_view
        self.progress = progress

        self.nw_view = NetworkView(self.db_view, self.progress)
        self.conflict_fixer = ConflictResolver(self.nw_view, self.progress)
        self.data_fixer = DataFixer(self.nw_view, self.progress)

    # ------------------------------------------------------------------------------
    #                                   CHANGE DETECTION
    # ------------------------------------------------------------------------------

    def changes(self, feature_types=None, change_types=None, bounds=None, limit=None):
        """
        The feature changes made in self's delta. Returns up to and includeing 'limit'
        changes within 'bounds'.

        Returns an ordered list of FeatureChange objects
        """

        if change_types is None:
            change_types = ["insert", "update", "delete"]

        changes = []
        cnt = 0
        for delta_rec in self.deltaRecs(
            feature_types=feature_types, change_types=change_types, bounds=bounds
        ):
            if limit and cnt >= limit:
                break
            cnt += 1

            base_rec = None
            if delta_rec.myw_change_type == "update":
                table = self.db_view.table(
                    delta_rec.feature_type
                )  # ENH: Get them all up front (for speed)
                base_rec = table._baseRec(delta_rec.id)

                if base_rec:
                    base_rec._view = self.db_view  # To allow getting display values
                else:
                    self.progress("warning", "Base record missing for", delta_rec)

            change = FeatureChange(delta_rec.myw_change_type, delta_rec, base_rec)
            changes.append(change)

        return changes

    def conflicts(self, bounds=None, categories=None):
        """
        Find conflicts in self's delta

        Optional BOUNDS is a polygon. Optional CATEGORIES is a list of names

        Returns a list of lists of MywConflict objects, keyed by feature type"""

        feature_types = self.orderedFeatureTypes(categories)
        change_types = ["insert", "update", "delete"]

        # Find conflicts
        conflicts = {}
        for feature_type in feature_types:
            with self.progress.operation("Checking", feature_type):
                table = self.db_view.table(feature_type)

                # Find changed records
                delta_recs = self.deltaRecs(
                    feature_types=[feature_type],
                    change_types=change_types,
                    bounds=bounds,
                )

                # Find conflicts
                ft_conflicts = {}
                for delta_rec in delta_recs:
                    conflict = self.conflictFor(table, delta_rec)

                    if conflict:
                        self.progress(0, conflict)
                        delta_rec._view = self.db_view
                        ft_conflicts[delta_rec._id] = conflict

                # Add to result
                if ft_conflicts:
                    conflicts[feature_type] = ft_conflicts

        return conflicts

    def validate(self, bounds=None, categories=None, max_errors=None):
        """
        Check integrity of objects in self's delta

        Returns a list of IntegrityError objects

        MAX_ERRORS - if specified will stop after that many errors found"""

        self.progress(6, "Checking", self.db_view)
        validator = DataValidator(self.db_view, progress=self.progress)

        feature_types = self.orderedFeatureTypes(categories)

        for rec in self.deltaRecs(bounds=bounds, feature_types=feature_types):

            validator.check(rec)

            if max_errors and len(validator.errors) >= max_errors:
                break

        return validator.errors

    # ------------------------------------------------------------------------------
    #                                   MERGING
    # ------------------------------------------------------------------------------

    def merge(self):
        """
        Auto-resolve conflicts and integrity errors in self's delta

        Returns a list of FeatureChange objects"""

        self.progress(2, "Merging", self.db_view)

        changes = {}
        self.fixConflicts(changes)
        self.fixGeoms(changes)

        return changes.values()

    def mergeRec(self, delta_rec):
        """
        Auto-resolve conflicts and integrity errors for DELTA_REC

        Returns dict with conflict and change"""

        table = self.db_view.table(delta_rec.feature_type, versioned_only=True)
        conflict = self.conflictFor(table, delta_rec)

        changes = {}
        if conflict:
            self.conflict_fixer.fixConflict(conflict, changes)

        rec = self.data_fixer.fixGeom(delta_rec, changes)
        table.update(rec)

        change = changes[delta_rec._urn()] if changes else None

        return {"conflict": conflict, "change": change}

    def fixConflicts(self, changes):
        """
        Auto-resolve conflicts in self's delta

        Updates CHANGES (a set of FeatureChange objects)"""

        with self.progress.operation("Fixing conflicts in", self.db_view):

            for feature_type, conflicts in self.conflicts().items():
                for id, conflict in conflicts.items():
                    self.conflict_fixer.fixConflict(conflict, changes)

    def fixGeoms(self, changes):
        """
        Update derived geometries on objects in self's delta

        Updates CHANGES (a set of FeatureChange objects)"""

        with self.progress.operation("Fixing geometries in", self.db_view):

            # Update routes
            for rec in self.deltaRecs(self.nw_view.routes):
                self.data_fixer.fixRouteGeom(rec, changes)

            # Update equipment
            for rec in self.deltaRecs(self.nw_view.equips):
                self.data_fixer.fixEquipGeom(rec, changes)

            # Update conduits
            extra_conduit_run_urns = set()
            for rec in self.deltaRecs(self.nw_view.conduits):
                if self.data_fixer.fixConduitGeom(rec, changes):
                    if "conduit_run" in rec._descriptor.fields and rec.conduit_run:
                        extra_conduit_run_urns.add(rec.conduit_run)

            # Update conduit runs
            for rec in self.deltaRecs(self.nw_view.conduit_runs):
                self.data_fixer.fixConduitRunGeom(rec, changes)
                extra_conduit_run_urns.discard(rec._urn())
            # ENH: Deal with extra_cables_urns (requires clone into view)

            # Update cable segments
            extra_cable_urns = set()
            for rec in self.deltaRecs(self.nw_view.segments):
                if self.data_fixer.fixSegmentGeom(rec, changes):
                    extra_cable_urns.add(rec.cable)

            # Update cables
            for rec in self.deltaRecs(self.nw_view.cables):
                self.data_fixer.fixCableGeom(rec, changes)
                extra_cable_urns.discard(rec._urn())
            # ENH: Deal with extra_cable_urns (requires clone into view)

            # Update connections
            for rec in self.deltaRecs(self.nw_view.connections):
                self.data_fixer.fixConnectionGeom(rec, changes)

            # Update circuits
            for rec in self.deltaRecs(self.nw_view.circuits):
                self.data_fixer.fixCircuitGeom(rec, changes)

            # Update line of count
            for rec in self.deltaRecs(self.nw_view.line_of_counts):
                self.data_fixer.fixLineOfCountGeom(rec, changes)

            self.progress(2, "Changes:", len(changes))

        return changes

    def revert(self, delta_rec):
        """
        Return delta_rec to master state
        """

        return self.conflict_fixer.revert(delta_rec)

    def rebase(self, delta_rec):
        """
        Update base record of DELTA_REC to match current master
        """

        return self.conflict_fixer.rebase(delta_rec)

    # ------------------------------------------------------------------------------
    #                                  HELPERS
    # ------------------------------------------------------------------------------

    def deltaRecs(self, feature_types=None, change_types=None, bounds=None):
        """
        The delta records of current view
        """
        # ENH: Replace by protocol on MywVersionedFeatureView
        # ENH: Return detached records?

        if change_types is None:
            change_types = ["insert", "update"]

        # Deal with defaults
        if feature_types is None:
            feature_types = self.orderedFeatureTypes()

        # For each feature type ..
        for feature_type in feature_types:
            tab = self.db_view.table(feature_type)

            # Build query (ordering to get inserts first)
            recs = tab._delta_recs.order_by(tab.delta_model._key_column().desc())
            if change_types:
                recs = recs.filter(tab.delta_model.myw_change_type.in_(change_types))

            if bounds:
                primary_geom_name = tab.descriptor.primary_geom_name

                if primary_geom_name is not None:
                    geom_filter = getattr(tab.delta_model, primary_geom_name).st_intersects(
                        bounds.ewkt()
                    )
                    recs = recs.filter(geom_filter)

            # Yield records
            for rec in recs:
                rec._view = self.db_view
                yield rec

    def orderedFeatureTypes(self, categories=None):
        """
        Versioned feature types, in 'top down' order

        If optional CATEGORIES is provided, return only types from those categories"""

        # ENH: Move to network view

        # Mapping from category name -> nw_view property
        name_mappings = {"structures": "structs", "equipment": "equips"}

        # Case: Categories
        # ENH: Filter result instead
        if categories:
            feature_types = []
            for category in categories:
                prop_name = name_mappings.get(category, category)
                feature_types += getattr(self.nw_view, prop_name)

            return feature_types

        # Case: All
        all_feature_types = self.db_view.db.dd.featureTypes("myworld", versioned_only=True)

        user_feature_types = (
            list(self.nw_view.structs)
            + list(self.nw_view.routes)
            + list(self.nw_view.equips)
            + list(self.nw_view.conduits)
            + list(self.nw_view.cables)
            + list(self.nw_view.connections)
            + list(self.nw_view.circuits)
            + list(self.nw_view.line_of_counts)
        )

        int_feature_types = self.nw_view.conduit_runs + list(self.nw_view.segments)

        custom_feature_types = sorted(
            list(set(all_feature_types) - set(user_feature_types) - set(int_feature_types))
        )

        return custom_feature_types + user_feature_types + int_feature_types

    def conflictFor(self, table, delta_rec):
        """
        Conflict info for delta_rec (if any)

        Returns a Conflict (or None)"""

        # ENH: Remove need for custom conflict class and get rid of this

        conflict = table.conflictFor(delta_rec)

        if not conflict:
            return None

        return Conflict(
            conflict.master_change,
            conflict.delta_rec,
            conflict.master_rec,
            conflict.base_rec,
        )

    def bounds(self):
        """
        Return convex hull around all changes in the delta
        For performance, constructs a database query against the delta geometry index tables
        """

        index_tables = [
            "delta_geo_world_point",
            "delta_geo_world_linestring",
            "delta_geo_world_polygon",
        ]

        models = self._getModels(Session, index_tables)

        from geoalchemy2 import Geometry
        from sqlalchemy import cast

        def make_query(model):
            query = Session.query(model)
            return query.filter(model.delta == self.db_view.delta).with_entities(
                cast(model.the_geom, Geometry)
            )

        base_queries = map(make_query, models.values())
        union_query = Query.union_all(*base_queries)
        subquery = union_query.subquery()

        hull_func = func.ST_GeogFromWKB(
            func.ST_convexhull(func.ST_collect(subquery.c.myw_delta_geo_world_point_the_geom))
        )
        results = Session.query(func.ST_astext(func.ST_buffer(hull_func, literal(10)))).select_from(
            subquery
        )

        bounds_geom = results.first()[0]

        return {"geometry": MywGeometry.decode(bounds_geom)} if bounds_geom else {}

    def _getModels(self, Session, tablenames):
        """
        Returns dictionary of record 'exemplars' for the geom index tables
        """
        # ENH: Get via driver

        import warnings
        from sqlalchemy import literal, Column, exists
        from myworldapp.core.server.models.myw_dd_feature import MywDDFeature
        from myworldapp.core.server.models.base import ModelBase, MywModelMixin
        from geoalchemy2 import Geometry

        # Suppressing SQLAlchemy warnings about geographic indexes
        with warnings.catch_warnings():
            from sqlalchemy import exc as sa_exc

            warnings.simplefilter("ignore", category=sa_exc.SAWarning)  # ENH: Should it be SA.exc?

            models = {}

            for a_tablename in tablenames:

                # Build a new table class
                model = type(
                    a_tablename,
                    (ModelBase, MywModelMixin),
                    dict(
                        __tablename__=Session.myw_db_driver.dbNameFor("myw", a_tablename),
                        __table_args__={
                            "schema": Session.myw_db_driver.dbNameFor("myw"),
                            "autoload": True,
                            "extend_existing": True,
                            "autoload_with": Session.bind,
                        },
                        the_geom=Column(Geometry(srid=4326)),
                    ),
                )

                # Add it to the dictionary
                models[a_tablename] = model

            return models
